include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td

```diff
@@ -129,6 +129,21 @@ def InstanceOp : HardwareDeclOp<"instance", [
     /// the new InstanceOp to the same location.
     InstanceOp cloneAndInsertPorts(ArrayRef<std::pair<unsigned, PortInfo>> ports);
 
+    //===------------------------------------------------------------------===//
+    // Instance graph methods
+    //===------------------------------------------------------------------===//
+
+    // Quick lookup of the referenced module using the instance graph.
+    template <typename T = ::circt::igraph::ModuleOpInterface>
+    T getReferencedModuleByGraph(::circt::igraph::InstanceGraph &instanceGraph) {
+      auto moduleNameAttr = getModuleNameAttr().getAttr();
+      auto *node =  instanceGraph.lookup(moduleNameAttr);
+      if (!node)
+        return nullptr;
+      Operation *moduleOp = node->getModule();
+      return dyn_cast_or_null<T>(moduleOp);
+    }
+
```

Comment:
Question for you (as I don't know the answer): there is a compile-time and code-size cost to doing this as a template vs. not (and pushing the casts into the use site). How do you balance this and is this something to worry about?

A slightly different factoring would use one common, non-templated member function (possibly private), and have the templated one call the non-templated one, adding a cast.

Nit: consider `getReferencedModule` or `getModule` if this is getting renamed.

Reply:
getReferencedModule is a better name.  The version which looks up a module without a symbol table or instancegraph should not exist.

I also would drop the template as @seldridge suggests.

Reply:
I take back my comment somewhat.  Every caller is casting anyway.  If we are really worried about the cost here, one could specialize the ModuleOpInterface version, and then write the generic in terms of that.  But that's saving very little and adding a call.